services/QuillLMS/spec/models/app_setting_spec.rb

```diff
@@ -0,0 +1,117 @@
+    end
+
+    context 'AppSetting does not exist' do
+      it 'should return false' do 
+        expect(AppSetting.enabled?(name: 'nemo', user: user)).to be false
+      end
+    end
+
+    context 'User and AppSetting exist' do 
+      it 'should invoke enabled_for_user?' do 
+        expect(AppSetting.enabled?(name: 'egosum', user: user)).to be true
+      end
+    end
+  end
+
+  describe '#all_enabled_for_user' do 
+    let(:user) { create(:user) }
+    it 'should return values for globally all enabled app settings' do 
+      create(:app_setting, name: 'enabled1', enabled: true, percent_active: 100)
+      create(:app_setting, name: 'enabled2', enabled: true, percent_active: 100)
+      create(:app_setting)
+      create(:app_setting)
+
+      expected = {
+        'enabled1' => true,
+        'enabled2' => true
+      }
+
+      expect(AppSetting.all_enabled_for_user(user)).to eq expected
+    end
+  end
+
+  describe '#enabled_for_user?' do 
+    context 'override is false' do 
+      let(:as1) { create(:app_setting, percent_active: 40, enabled: false) }
+
+      it 'should return false' do 
+        u1 = create(:user)
+        expect(as1.enabled_for_user?(u1)).to be false
+      end      
+
+    end
+
+    context 'override is true, admin is true' do 
+      let(:as1) { create(:app_setting, percent_active: 0, enabled: true, enabled_for_admins: true) }
+
+      it 'should return true when user is admin' do 
+        allow_any_instance_of(User).to receive(:admin?) { true }
+        u1 = create(:user)
+        expect(as1.enabled_for_user?(u1)).to be true
+      end
+    
+      it 'should return true for non-admin user when user is in whitelist' do 
+        allow_any_instance_of(User).to receive(:admin?) { false } 
+        u1 = create(:user)
+        as1.user_ids_allow_list = [u1.id]
+        
+        expect(as1.enabled_for_user?(u1)).to be true
+      end    
+
+      it 'should return false when user does not meet enabled criteria' do 
+        u1 = create(:user)
+        expect(as1.enabled_for_user?(u1)).to be false
+      end
+    end
+
+  end
+
+  describe '#user_in_rollout_bucket?' do 
+    let(:random_numbers) { (1..1000).map(&:to_i) }
+
+    context 'statistical sanity test for hashing algorithm' do 
+      it 'when percent_active is 50, users should be evenishly distributed' do 
+        app_setting = create(:app_setting, name: 'lorem1', enabled: true, percent_active: 50)
+        result = random_numbers.map {|r| app_setting.user_in_rollout_bucket?(r) }
+        expect(result.filter{|r| r}.count).to eq 482
+      end
+
+      it 'when percent_active is low, few users should be in bucket' do
+        app_setting = create(:app_setting, name: 'lorem2', enabled: true, percent_active: 10)
+        result = random_numbers.map {|r| app_setting.user_in_rollout_bucket?(r) }
+        expect(result.filter{|r| r}.count).to eq 97
+      end
+
+      it 'when percent_active is high, many users should be in bucket' do
```

Comment:
the overall file takes .812 s with the 1000-count sanity checks. I've reduced the count to 100, and now the overall file takes .629 s.  I started with count=1000 because I wanted to prove to myself that the algorithm worked. Now that we know it works, a smaller counter (+bigger skew) is acceptable.